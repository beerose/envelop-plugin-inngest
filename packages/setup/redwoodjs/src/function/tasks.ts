import path from 'path';
import camelcase from 'camelcase';
import execa from 'execa';
import fs from 'fs-extra';
import humanize from 'humanize-string';
import * as jscodeshift from 'jscodeshift/src/Runner';
import { Listr } from 'listr2';
import template from 'lodash.template';
import { paramCase } from 'param-case';
import { getPaths, writeFile } from '@redwoodjs/cli-helpers';
import type { ExistingFiles } from '@redwoodjs/cli-helpers';
import type { SetupInngestFunctionOptions } from './command';

export interface SetupFunctionTasksOptions extends SetupInngestFunctionOptions {}

const getNamesForFile = (name: string) => {
  const functionName = camelcase(name);
  const humanizedName = humanize(name);
  const eventName = paramCase(name).replace(/-/g, '.');

  return { functionName, humanizedName, eventName };
};

const SRC_INNGEST_PATH = path.join(getPaths().api.src, 'inngest');

const renderFunctionTemplate = (name: string, type: string) => {
  const { eventName, functionName, humanizedName } = getNamesForFile(name);

  const compiled = template(
    fs
      .readFileSync(
        path.resolve(__dirname, '..', '..', 'templates', 'function', `${type}.ts.template`),
        'utf-8',
      )
      .toString(),
  );

  const rendered = compiled({ name, eventName, functionName, humanizedName });

  return { filename: functionName, rendered };
};

const writeFunctionFile = (filename: string, rendered: string, existingFiles: ExistingFiles) => {
  writeFile(path.join(SRC_INNGEST_PATH, `${filename}.ts`), rendered, {
    existingFiles,
  });
};

export const tasks = (options: SetupFunctionTasksOptions) => {
  const existingFiles = options.force ? 'OVERWRITE' : 'FAIL';

  return new Listr(
    [
      {
        title: `Create a ${options.type} Inngest function named ${options.name} ...`,
        task: () => {
          const { filename, rendered } = renderFunctionTemplate(options.name, options.type);
          writeFunctionFile(filename, rendered, existingFiles);
        },
      },
      {
        title: `Modify the Inngest handler to register the new ${options.type} Inngest function function  named ${options.name}...`,
        task: async () => {
          const { functionName } = getNamesForFile(options.name);

          const SRC_INNGEST_HANDLER_FILE = path.join(getPaths().api.functions, 'inngest.ts');
          const SRC_INNGEST_CODEMOD_FILE = path.join(
            __dirname,
            '..',
            'add-function-to-inngest-handler-codemod.js',
          );

          const defaultJscodeshiftOpts = {
            verbose: 0,
            dry: false,
            print: false,
            babel: false,
            extensions: 'js',
            ignorePattern: '**/node_modules/**',
            ignoreConfig: [],
            runInBand: false,
            silent: false,
            parser: 'ts',
            parserConfig: {},
            failOnError: true,
            stdin: false,
          };

          try {
            await jscodeshift.run(SRC_INNGEST_CODEMOD_FILE, [SRC_INNGEST_HANDLER_FILE], {
              ...defaultJscodeshiftOpts,
              functionName,
            });
          } catch (e: any) {
            // eslint-disable-next-line no-console
            console.error('Failed to modify the GraphQL handler', e.message);
          }
        },
      },
      {
        title: `Lint and Prettify added files ...`,
        task: async () => {
          const SRC_INNGEST_HANDLER_FILE = path.join(getPaths().api.functions, 'inngest.ts');

          execa.commandSync(
            `yarn rw lint --fix ${SRC_INNGEST_HANDLER_FILE}`,
            // eslint-disable-next-line dot-notation
            process.env['RWJS_CWD']
              ? {
                  // eslint-disable-next-line dot-notation
                  cwd: process.env['RWJS_CWD'],
                }
              : {},
          );
        },
      },
    ],
    { rendererOptions: {} },
  );
};
