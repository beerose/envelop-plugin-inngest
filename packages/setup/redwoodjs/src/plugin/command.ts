import type Yargs from 'yargs';

interface BaseOptions {
  cwd: string | undefined;
}

export interface ForceOptions extends BaseOptions {
  force: boolean;
}

export const description = 'Set up Inngest plugin';

export const builder = (yargs: Yargs.Argv<BaseOptions>) => {
  return yargs.option('force', {
    alias: 'f',
    default: true,
    description: 'Overwrite existing files',
    type: 'boolean',
  });
};

export const handler = async (options: ForceOptions) => {
  // setup plugin
  // eslint-disable-next-line no-console
  console.debug('setup plugin', options);
  const { handler } = await import('./handler');
  return handler(options);
};
