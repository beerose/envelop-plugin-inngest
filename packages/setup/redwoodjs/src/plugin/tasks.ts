import path from 'path';
import execa from 'execa';
import fs from 'fs-extra';
import * as jscodeshift from 'jscodeshift/src/Runner';
import { Listr } from 'listr2';
import { getPaths, writeFile } from '@redwoodjs/cli-helpers';
import type { ForceOptions } from './command';

export interface SetupPluginTasksOptions extends ForceOptions {}

interface PackageJson {
  scripts?: Record<string, string>;
}

export const tasks = (options: SetupPluginTasksOptions) => {
  const PACKAGE_JSON_PATH = path.join(getPaths().base, 'package.json');
  const SRC_INNGEST_PATH = path.join(getPaths().api.src, 'jobs', 'inngest');
  const SRC_LIB_PATH = path.join(getPaths().api.lib);
  const SRC_PLUGINS_PATH = path.join(getPaths().api.src, 'plugins');
  const SRC_GRAPHQL_FUNCTION_FILE = path.join(getPaths().api.functions, 'graphql.ts');
  const SRC_INNGEST_CODEMOD_FILE = path.join(__dirname, '..', 'use-inngest-codemod.js');

  const commandPaths = {
    PACKAGE_JSON_PATH,
    SRC_GRAPHQL_FUNCTION_FILE,
    SRC_INNGEST_CODEMOD_FILE,
    SRC_INNGEST_PATH,
    SRC_LIB_PATH,
    SRC_PLUGINS_PATH,
  };
  const existingFiles = options.force ? 'OVERWRITE' : 'FAIL';

  return new Listr(
    [
      {
        title: 'Install inngest packages ...',
        task: () => {
          execa.commandSync(
            'yarn workspace api add envelop-plugin-inngest',
            // eslint-disable-next-line dot-notation
            process.env['RWJS_CWD']
              ? {
                  // eslint-disable-next-line dot-notation
                  cwd: process.env['RWJS_CWD'],
                }
              : {},
          );
        },
      },

      {
        title: 'Configure inngest ...',
        task: () => {
          configureInngestTask({ commandPaths, existingFiles });
        },
      },
      {
        title: 'Add the Inngest GraphQL plugin ...',
        task: () => {
          addInngestGraphQLPluginTask({ commandPaths, existingFiles });
        },
      },
      {
        title: 'Add inngest helloWorld example ...',
        task: () => {
          addInngestHelloWorldExampleTask({ commandPaths, existingFiles });
        },
      },
      {
        title: 'Modify the GraphQL handler to the useInngest plugin ...',
        task: async () => {
          modifyGraphQLHandlerTask({ commandPaths });
        },
      },
      {
        title: 'Add inngest dev script to package.json',
        task: async () => {
          addScriptToPackageJsonTask({ commandPaths });
        },
      },
      {
        title: `Lint and Prettify added files ...`,
        task: async () => {
          const SRC_GRAPHQL_FUNCTION_FILE = path.join(getPaths().api.functions, 'graphql.ts');

          execa.commandSync(
            `yarn rw lint --fix ${SRC_GRAPHQL_FUNCTION_FILE}`,
            // eslint-disable-next-line dot-notation
            process.env['RWJS_CWD']
              ? {
                  // eslint-disable-next-line dot-notation
                  cwd: process.env['RWJS_CWD'],
                }
              : {},
          );
        },
      },
    ],
    { rendererOptions: {} },
  );
};

export const addScriptToPackageJsonTask = ({
  commandPaths,
}: {
  commandPaths: Record<string, string>;
}) => {
  const packageJson = JSON.parse(
    fs.readFileSync(commandPaths.PACKAGE_JSON_PATH, 'utf-8'),
  ) as PackageJson;

  if (!packageJson.scripts) {
    packageJson.scripts = {};
  }

  packageJson.scripts['inngest:dev'] =
    'npx inngest-cli@latest dev -u http://localhost:8911/inngest';

  writeFile(commandPaths.PACKAGE_JSON_PATH, JSON.stringify(packageJson, null, 2), {
    existingFiles: 'OVERWRITE',
  });
};

export const configureInngestTask = ({
  commandPaths,
  existingFiles,
}: {
  commandPaths: Record<string, string>;
  existingFiles: 'OVERWRITE' | 'FAIL';
}) => {
  // save inngest handler function in api functions
  const inngestServerFunctionTemplate = fs.readFileSync(
    path.resolve(__dirname, '..', '..', 'templates', 'plugin', 'inngest.ts.template'),
    'utf-8',
  );

  writeFile(path.join(getPaths().api.functions, 'inngest.ts'), inngestServerFunctionTemplate, {
    existingFiles,
  });

  // save inngest client to a api lib
  fs.ensureDirSync(commandPaths.SRC_LIB_PATH);

  const inngestClientTemplate = fs.readFileSync(
    path.resolve(__dirname, '..', '..', 'templates', 'plugin', 'client.ts.template'),
    'utf-8',
  );

  writeFile(path.join(commandPaths.SRC_LIB_PATH, 'inngest.ts'), inngestClientTemplate, {
    existingFiles: 'OVERWRITE',
  });
};

const addInngestGraphQLPluginTask = ({
  commandPaths,
  existingFiles,
}: {
  commandPaths: Record<string, string>;
  existingFiles: 'OVERWRITE' | 'FAIL';
}) => {
  // save inngest plugin to a new plugins folder
  fs.ensureDirSync(commandPaths.SRC_PLUGINS_PATH);

  const inngestPluginTemplate = fs.readFileSync(
    path.resolve(__dirname, '..', '..', 'templates', 'plugin', 'plugin.ts.template'),
    'utf-8',
  );

  writeFile(path.join(commandPaths.SRC_PLUGINS_PATH, 'useInngest.ts'), inngestPluginTemplate, {
    existingFiles,
  });
};

const addInngestHelloWorldExampleTask = ({
  commandPaths,
  existingFiles,
}: {
  commandPaths: Record<string, string>;
  existingFiles: 'OVERWRITE' | 'FAIL';
}) => {
  // save example inngest functions in the inngest folder
  fs.ensureDirSync(commandPaths.SRC_INNGEST_PATH);

  const inngestHelloWorldTemplate = fs.readFileSync(
    path.resolve(__dirname, '..', '..', 'templates', 'plugin', 'helloWorld.ts.template'),
    'utf-8',
  );

  return writeFile(
    path.join(commandPaths.SRC_INNGEST_PATH, 'helloWorld.ts'),
    inngestHelloWorldTemplate,
    {
      existingFiles,
    },
  );
};

export const modifyGraphQLHandlerTask = async ({
  commandPaths,
}: {
  commandPaths: Record<string, string>;
}) => {
  const defaultJscodeshiftOpts = {
    verbose: 0,
    dry: false,
    print: false,
    babel: false,
    extensions: 'js',
    ignorePattern: '**/node_modules/**',
    ignoreConfig: [],
    runInBand: false,
    silent: true,
    parser: 'ts',
    parserConfig: {},
    failOnError: true,
    stdin: false,
  };

  try {
    await jscodeshift.run(
      commandPaths.SRC_INNGEST_CODEMOD_FILE,
      [commandPaths.SRC_GRAPHQL_FUNCTION_FILE],
      {
        ...defaultJscodeshiftOpts,
      },
    );
  } catch (e: any) {
    // eslint-disable-next-line no-console
    console.error('Failed to modify the GraphQL handler', e.message);
  }
};
