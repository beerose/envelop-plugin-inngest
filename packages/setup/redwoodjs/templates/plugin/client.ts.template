import { Inngest } from 'inngest'

export const INNGEST_APP_NAME = 'Redwood_Inngest'

// inngest client to be used in services and functions
export const inngest = new Inngest({
  name: INNGEST_APP_NAME,
  eventKey: 'Redwood',
})
