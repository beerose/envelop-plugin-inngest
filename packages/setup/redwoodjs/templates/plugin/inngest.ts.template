import { serve } from 'inngest/redwood'

import { INNGEST_APP_NAME } from 'src/inngest/client'
import helloWorld from 'src/inngest/helloWorld'

export const handler = serve(INNGEST_APP_NAME, [helloWorld], {
  servePath: '/.redwood/functions/inngest',
  signingKey: 'Redwood',
})
