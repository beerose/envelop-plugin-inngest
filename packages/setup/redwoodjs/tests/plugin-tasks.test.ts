import path from 'path';
import { addScriptToPackageJsonTask, tasks } from '../src/plugin/tasks';

let f = '';

jest.mock('@redwoodjs/cli-helpers', () => {
  const path = require('path');
  const __dirname = path.resolve();

  return {
    getPaths: () => ({
      api: {
        src: path.join('__testfixtures__', 'plugin', 'create-redwood-app/template/api/src'),
        functions: path.join(
          '__testfixtures__',
          'plugin',
          'create-redwood-app/template/api/src/functions',
        ),
        lib: path.join('__testfixtures__', 'plugin', 'create-redwood-app/template/api/src/lib'),
      },
      web: { src: '' },
      base: path.join('__testfixtures__', 'plugin', 'create-redwood-app/template'),
    }),
    writeFile: (target: string, contents: string) => {
      f = contents;
    },
  };
});

// jest.mock('../../lib/project', () => ({
//   isTypeScriptProject: () => true,
// }));

describe('Plugin tasks', () => {
  it('has tasks', () => {
    expect(tasks).toBeDefined();
  });

  it('has addScriptToPackageJsonTask', () => {
    expect(addScriptToPackageJsonTask).toBeDefined();
    const commandPaths = {
      PACKAGE_JSON_PATH: path.join(__dirname, '__testfixtures__', 'plugin', 'test.package.json'),
    };

    //  packages/setup/redwoodjs/tests/__testfixtures__/plugin/package.json
    addScriptToPackageJsonTask({ commandPaths });

    expect(f).toMatchSnapshot();
  });
});
