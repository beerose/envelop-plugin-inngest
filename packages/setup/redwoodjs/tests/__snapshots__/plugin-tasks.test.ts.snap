// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Plugin tasks has addInngestGraphQLPluginTask 1`] = `
"// Setup the envelop plugin used in RedwoodJS GraphQLHandler

import { useInngest } from 'envelop-plugin-inngest'
import type {
  BuildUserContextFunction,
  InngestUserContextOptions,
} from 'envelop-plugin-inngest'

import { inngest } from 'src/lib/inngest'

// Includes the user context for each authenticated GraphQL request
export const userContext: BuildUserContextFunction = (
  options: InngestUserContextOptions
) => {
  const currentUser = options?.params?.args?.contextValue?.currentUser

  if (currentUser) {
    return { id: currentUser.id }
  }

  return {}
}

// For a complete list of inngest plugin configuration options, see: https://github.com/inngest/envelop-plugin-inngest/blob/main/packages/plugins/inngest/README.md
export const inngestPlugin = useInngest({
  inngestClient: inngest,
  buildUserContextFunction: userContext,
})
"
`;

exports[`Plugin tasks has addInngestHelloWorldExampleTask 1`] = `
"import { inngest } from 'src/lib/inngest'

const helloWorld = inngest.createFunction(
  { name: 'Hello World' },
  { event: 'test/hello-world' },
  async ({ event, step }) => {
    await step.sleep('1s')

    return {
      event,
      body: \`Hello, \${event.data?.name ? event.data.name : 'World'}!\`,
    }
  }
)

export default helloWorld
"
`;

exports[`Plugin tasks has addScriptToPackageJsonTask 1`] = `
"{
  "private": true,
  "workspaces": {
    "packages": [
      "api",
      "web"
    ]
  },
  "engines": {
    "node": ">=16.19 <=18.x",
    "yarn": ">=1.15"
  },
  "devDependencies": {
    "@redwoodjs/core": "4.4.2",
    "inngest-setup-redwoodjs": "0.3.0",
    "prettier-plugin-tailwindcss": "^0.2.7"
  },
  "eslintConfig": {
    "extends": "@redwoodjs/eslint-config",
    "root": true
  },
  "prisma": {
    "seed": "yarn rw exec seed"
  },
  "scripts": {
    "inngest:dev": "npx inngest-cli@latest dev -u http://localhost:8911/inngest"
  }
}"
`;

exports[`Plugin tasks has configureInngestTask 1`] = `
"import { Inngest } from 'inngest'

export const INNGEST_APP_NAME = 'Redwood_Inngest'

export const inngest = new Inngest({
  /**
   * The name of this instance, most commonly the name of the application it
   * resides in.
   */
  name: INNGEST_APP_NAME,
  /**
   * Inngest event key, used to send events to Inngest Cloud. If not provided,
   * will search for the \`INNGEST_EVENT_KEY\` environment variable. If neither
   * can be found, however, a warning will be shown and any attempts to send
   * events will throw an error.
   */
  eventKey: 'YOUR_INNGEST_EVENT_KEY',
})
"
`;

exports[`Plugin tasks has updateTomlConfig 1`] = `
"
[experimental.jobs.inngest.plugin]"
`;
